{% extends "layout.html" %}
{% block breadcrumbs %}
<li>{{ screen.name }}</li>
<li class="active">quality control</li>
{% endblock %}
{% block content %}
<div class="row">
    <div class="col-md-5">
        {% if fastqc %}
        <div class="row">
            <div class="col-md-12">
                <h4>GC content</h4>
                <div class="text-center">
                    <div id="gc-content" class="plot" data-toggle="popover" data-desc="Per sample distribution of mean GC content of reads. The mean should be between 40 and 50."></div>
                </div>
            </div>
        </div>
        {% endif %}
    </div>
    <div class="col-md-7">
        {% if fastqc %}
        <div class="row">
            <div class="col-md-12">
                <h4>Base quality</h4>
                <div class="text-center">
                    <div id="base-quality" class="plot" data-toggle="popover" data-desc="Median base quality per-read base for each sample."></div>
                    <div id="seq-quality" class="plot" data-toggle="popover" data-placement="left" data-desc="Distribution of per-read mean base quality for each sample. The peak should be beyond 30."></div>
                </div>
            </div>
        </div>
        {% endif %}
    </div>
</div>
{% if mapstats %}
<div class="row">
    <div class="col-md-12">
        <h4>Mapping statistics</h4>
        <div class="text-center">
            <div id="mapped-unmapped" class="plot" data-toggle="popover" data-desc="Mapped and unmapped reads for each sample. Percentage of unmapped reads is printed on each bar."></div>
            <div id="zerocounts" class="plot" data-toggle="popover" data-desc="Per-sample number of missed gRNAs, i.e. gRNAs with zero mapped reads."></div>
            <div id="gini-index" class="plot" data-toggle="popover" data-placement="left" data-desc="Gini Index for each sample. This measures inequality between read counts. 1.0 means maximum inequality (e.g. all reads mapping to one gRNA), 0.0 corresponds to having the same read count for all gRNAs."></div>
        </div>
    </div>
</div>
{% endif %}
<div class="row">
    <div class="col-md-5">
        <div class="row">
            <div class="col-md-12">
                <h4>Normalized read count distributions</h4>
                <div class="text-center">
                    <div id="readcounts" class="plot" data-toggle="popover" data-desc="Distribution of normalized read counts for each sample."></div>
                    <div id="readcount-cdf" class="plot" data-toggle="popover" data-desc="Cumulative distribution function of normalized read counts for each sample."></div>
                </div>
            </div>
        </div>
        <div class="row">
            <div class="col-md-12">
                <h4>Sample correlation</h4>
                <div class="text-center">
                    <div id="correlation" class="plot" data-toggle="popover" data-desc="Correlation between normalized log10 read counts of samples."></div>
                </div>
            </div>
        </div>
    </div>
    <div class="col-md-7">
        <div class="row">
            <div class="col-md-12">
                <h4>Principal component analysis</h4>
                <div class="text-center">
                    <div id="pca-pc1-pc2" class="plot" data-toggle="popover" data-desc="Principal component analysis on normalized log10 read counts of samples. The axis labels show principal components and percentage of explained variants."></div>
                    <div id="pca-pc1-pc3" class="plot" data-toggle="popover" data-desc="Principal component analysis on normalized log10 read counts of samples. The axis labels show principal components and percentage of explained variants."></div>
                    <div id="pca-pc2-pc3" class="plot" data-toggle="popover" data-placement="left" data-desc="Principal component analysis on normalized log10 read counts of samples. The axis labels show principal components and percentage of explained variants."></div>
                </div>
            </div>
        </div>
    </div>
</div>

<script>
vg.parse.spec("/plt/normalization/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#readcounts");
});
vg.parse.spec("/plt/readcount_cdf/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#readcount-cdf");
});
vg.parse.spec("/plt/pca/{{ screen.name }}/1/2/0", function(chart) {
    vega_draw(chart, "#pca-pc1-pc2");
});
vg.parse.spec("/plt/pca/{{ screen.name }}/1/3/0", function(chart) {
    vega_draw(chart, "#pca-pc1-pc3");
});
vg.parse.spec("/plt/pca/{{ screen.name }}/2/3/1", function(chart) {
    vega_draw(chart, "#pca-pc2-pc3");
});
vg.parse.spec("/plt/correlation/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#correlation");
});

{% if fastqc %}
vg.parse.spec("/plt/gc_content/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#gc-content");
});
vg.parse.spec("/plt/base_quality/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#base-quality");
});
vg.parse.spec("/plt/seq_quality/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#seq-quality");
});
{% endif %}

{% if mapstats %}
vg.parse.spec("/plt/mapstats/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#mapped-unmapped");
});
vg.parse.spec("/plt/zerocounts/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#zerocounts");
});
vg.parse.spec("/plt/gini_index/{{ screen.name }}", function(chart) {
    vega_draw(chart, "#gini-index");
});
{% endif %}

</script>
{% endblock %}
